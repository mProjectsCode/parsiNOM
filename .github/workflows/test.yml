name: Test

on:
  push:
    branches: [master, main]
  pull_request:
    branches: [master, main]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest
      - working-directory: .
        run: bun install
      - working-directory: .
        run: bun run check
