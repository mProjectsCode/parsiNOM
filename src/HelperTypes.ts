import type { Parser } from './Parser';
import type { ParserContext } from './ParserContext';

export type STypeBase = unknown;

export interface ParseSuccess<SType extends STypeBase> {
	success: true;
	value: SType;
	furthest?: undefined;
	expected?: undefined;
}

export interface ParseFailure {
	success: false;
	value: undefined;
	furthest: number;
	expected: string[];
}

export type ParseResult<SType extends STypeBase> = ParseSuccess<SType> | ParseFailure;

export interface InternalParseFailure {
	success: false;
	value: undefined;
}

export type InternalParseResult<SType extends STypeBase> = ParseSuccess<SType> | InternalParseFailure;

export type ParseFunction<SType extends STypeBase> = (context: ParserContext) => InternalParseResult<SType>;

export interface ParsingRange {
	from: number;
	to: number;
}

export interface ParsingMarker<SType extends STypeBase> {
	value: SType;
	range: ParsingRange;
}

export interface NamedParsingMarker<SType extends STypeBase> {
	value: SType;
	name: string;
	range: ParsingRange;
}

export type LanguageDef<RuleNames extends object> = {
	[P in keyof RuleNames]: Parser<RuleNames[P]>;
};

export type LanguageRules<RuleNames extends object> = {
	[P in keyof RuleNames]: (language: LanguageDef<RuleNames>) => Parser<RuleNames[P]>;
};

/**
 * All parsers of a language. Returned by {@link P.createLanguage}.
 */
export type NomLanguage<RulesType extends object> = {
	[P in keyof RulesType]: Parser<RulesType[P]>;
};

/**
 * All parsers of a language. Returned by {@link P.createLanguage}.
 */
export type NomLanguageRules<RulesType extends object> = {
	readonly [P in keyof RulesType]: (language: NomLanguagePartial<RulesType>, ref: NomLanguageRef<RulesType>) => Parser<RulesType[P]>;
};

/**
 * All parsers of a language wrapped in {@link P.reference}.
 */
export type NomLanguageRef<RulesType extends object> = {
	[P in keyof RulesType]: ParserRef<RulesType[P]>;
};

/**
 * Parsers of a language that came before this parser. All other parsers will be undefined.
 */
export type NomLanguagePartial<RulesType extends object> = {
	[P in keyof RulesType]: Parser<RulesType[P]>;
};

export type ParserRef<SType extends STypeBase> = Parser<SType>;

export type TupleToUnion<T extends readonly unknown[]> = T[number];

export type DeParser<T extends Parser<unknown>> = T extends Parser<infer U> ? U : never;

export type DeParserArray<T extends readonly Parser<unknown>[]> = {
	[I in keyof T]: T[I] extends Parser<infer R> ? R : never;
};
