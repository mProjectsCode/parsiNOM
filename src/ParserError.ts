import type { ParseFailure } from './HelperTypes';

/**
 * Generate an error message string for a parse failure on a specific string.
 *
 * @param str the input string that the parse failure occurred on
 * @param parseFailure the failure to generate the error message for
 * @param verbose will underline the failure position in the input string, if set to true
 */
export function createParsingErrorMessage(str: string, parseFailure: ParseFailure, verbose: boolean): string {
	const { line, column } = mapIndexToLineColumn(str, parseFailure.furthest);

	const expectedMessage = `Expected ${parseFailure.expected.sort().join(' or ')}`;
	let message = `Parse Failure: ${expectedMessage} at index ${parseFailure.furthest}, line ${line}, column ${column}.`;

	if (verbose) {
		const lines = str.split('\n');
		const failedLine = lines[line - 1]; // line is a one based index

		const linePrefix = `${line} |   `;
		message += `\n\n${linePrefix}${failedLine}`;
		message += `\n${' '.repeat(column - 1 + linePrefix.length)}^ (${expectedMessage})`;
	}

	return message;
}

/**
 * Will be thrown if parsiNOM fails to parse an input string.
 */
export class ParsingError extends Error {
	constructor(str: string, parseFailure: ParseFailure) {
		super(createParsingErrorMessage(str, parseFailure, true));
	}
}

export function mapIndexToLineColumn(str: string, index: number): { line: number; column: number } {
	let line = 1;
	let column = 1;

	for (let i = 0; i < index && i < str.length; i++) {
		if (str[i] === '\n') {
			line++;
			column = 1;
		} else {
			column++;
		}
	}

	return { line, column };
}
