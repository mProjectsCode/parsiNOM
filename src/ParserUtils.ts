import type {ParsingPosition, ParsingRange} from './HelperTypes';
import { Parser } from './Parser';
import { P } from './ParsiNOM';

export class P_UTILS {
	/**
	 * Yields the current position in the input string.
	 */
	static position(): Parser<ParsingPosition> {
		return new Parser<ParsingPosition>(context => {
			return context.succeed(context.getPosition());
		});
	}

	/**
	 * Matches any single character except for eof.
	 * Yields the character
	 */
	static any(): Parser<string> {
		return new Parser<string>(context => {
			if (context.atEOF()) {
				return context.fail('any character');
			}
			return context.succeedOffset(1, context.input[context.position.index]);
		});
	}

	/**
	 * Matches the entire rest of the string until the end.
	 * Yields the rest of the string.
	 */
	static remaining(): Parser<string> {
		return new Parser<string>(context => {
			return context.succeedAt(context.input.length, context.input.slice(context.position.index));
		});
	}

	/**
	 * Matches the end of the input.
	 */
	static eof(): Parser<undefined> {
		return new Parser<undefined>(context => {
			if (!context.atEOF()) {
				return context.fail('eof');
			}
			return context.succeed(undefined);
		});
	}

	/**
	 * Matches a single digit.
	 */
	static digit(): Parser<string> {
		return P.regexp(/^[0-9]/).describe('a digit');
	}

	/**
	 * Matches multiple digits.
	 */
	static digits(): Parser<string> {
		return P.regexp(/^[0-9]+/).describe('multiple digits');
	}

	/**
	 * Matches a single ascii letter.
	 */
	static letter(): Parser<string> {
		return P.regexp(/^[a-z]/i).describe('a letter');
	}

	/**
	 * Matches multiple ascii letters.
	 */
	static letters(): Parser<string> {
		return P.regexp(/^[a-z]+/i).describe('multiple letters');
	}

	/**
	 * Matches a single unicode letter.
	 */
	static unicodeLetter(): Parser<string> {
		return P.regexp(/^\p{L}/iu).describe('a unicode letter');
	}

	/**
	 * Matches multiple unicode letters.
	 */
	static unicodeLetters(): Parser<string> {
		return P.regexp(/^\p{L}+/iu).describe('multiple unicode letters');
	}

	/**
	 * Matches a single alphanumeric unicode character.
	 */
	static unicodeAlphanumeric(): Parser<string> {
		return P.regexp(/^[\p{L}\p{N}]/iu).describe('a unicode alphanumeric character');
	}

	/**
	 * Matches multiple alphanumeric unicode characters.
	 */
	static unicodeAlphanumerics(): Parser<string> {
		return P.regexp(/^[\p{L}\p{N}]+/iu).describe('multiple unicode alphanumeric characters');
	}

	/**
	 * Matches as much whitespace as it can or no whitespace.
	 */
	static optionalWhitespace(): Parser<string> {
		return P.regexp(/^\s*/).describe('optional whitespace');
	}

	/**
	 * Matches multiple, at least one, whitespace.
	 */
	static whitespace(): Parser<string> {
		return P.regexp(/^\s+/).describe('whitespace');
	}

	/**
	 * Matches a `\r` character.
	 */
	static cr(): Parser<string> {
		return P.string('\r');
	}

	/**
	 * Matches a `\n` character.
	 */
	static lf(): Parser<string> {
		return P.string('\n');
	}

	/**
	 * Matches a `\r\n` character.
	 */
	static crlf(): Parser<string> {
		return P.string('\r\n');
	}

	/**
	 * Matches a newline character (either `\r\n`, `\n` or `\r`).
	 */
	static newline(): Parser<string> {
		return P.or(this.crlf(), this.lf(), this.cr()).describe('newline');
	}

	/**
	 * Matches as many operators as it can and then a singe operand. Then it will reduce the list of operators with the combine function.
	 *
	 * @param operatorsParser
	 * @param nextParser
	 * @param combine
	 */
	static prefix<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		nextParser: Parser<OtherSType>,
		combine: (a: OperatorSType, b: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(prefixes: OperatorSType[], x: OtherSType) => {
				return prefixes.reduce<OtherSType | ReturnSType>((acc, y) => combine(y, acc), x);
			},
			operatorsParser.many(),
			nextParser,
		);
	}

	/**
	 * Matches a singe operand and then as many operators as it can. Then it will reduce the list of operators with the combine function.
	 *
	 * @param operatorsParser
	 * @param nextParser
	 * @param combine
	 */
	static postfix<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		nextParser: Parser<OtherSType>,
		combine: (a: OperatorSType, b: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(x: OtherSType, postfixes: OperatorSType[]) => {
				return postfixes.reduce<OtherSType | ReturnSType>((acc, y) => combine(y, acc), x);
			},
			nextParser,
			operatorsParser.many(),
		);
	}

	/**
	 * Matches a right associative binary operation.
	 *
	 * @param operatorsParser
	 * @param operandParser
	 * @param combine
	 */
	static binaryRight<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		operandParser: Parser<OtherSType>,
		combine: (a: OtherSType, b: OperatorSType, c: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(others: [OtherSType, OperatorSType][], last: OtherSType) => {
				return others.reverse().reduce<OtherSType | ReturnSType>((acc, y) => {
					const [operand, operator] = y;
					return combine(operand, operator, acc);
				}, last);
			},
			P.sequence(operandParser, operatorsParser.trim(this.optionalWhitespace())).many(),
			operandParser,
		);
	}

	/**
	 * Matches a left associative binary operation.
	 *
	 * @param operatorsParser
	 * @param operandParser
	 * @param combine
	 */
	static binaryLeft<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		operandParser: Parser<OtherSType>,
		combine: (a: OtherSType | ReturnSType, b: OperatorSType, c: OtherSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(first: OtherSType, others: [OperatorSType, OtherSType][]) => {
				return others.reduce<OtherSType | ReturnSType>((acc, y) => {
					const [operator, operand] = y;
					return combine(acc, operator, operand);
				}, first);
			},
			operandParser,
			P.sequence(operatorsParser.trim(this.optionalWhitespace()), operandParser).many(),
		);
	}

	/**
	 * Matches a right associative binary operation and also passes a parsing range to the combine function.
	 *
	 * @param operatorsParser
	 * @param operandParser
	 * @param combine
	 */
	static binaryRightRange<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		operandParser: Parser<OtherSType>,
		combine: (range: ParsingRange, a: OtherSType, b: OperatorSType, c: OtherSType | ReturnSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(others: [ParsingPosition, OtherSType, OperatorSType][], last: OtherSType, to: ParsingPosition) => {
				return others.reverse().reduce<OtherSType | ReturnSType>((acc, y) => {
					const [from, operand, operator] = y;
					return combine({ from, to }, operand, operator, acc);
				}, last);
			},
			P.sequence(P_UTILS.position(), operandParser, operatorsParser.trim(this.optionalWhitespace())).many(),
			operandParser,
			P_UTILS.position(),
		);
	}

	/**
	 * Matches a left associative binary operation and also passes a parsing range to the combine function.
	 *
	 * @param operatorsParser
	 * @param operandParser
	 * @param combine
	 */
	static binaryLeftRange<OperatorSType, OtherSType, ReturnSType>(
		operatorsParser: Parser<OperatorSType>,
		operandParser: Parser<OtherSType>,
		combine: (range: ParsingRange, a: OtherSType | ReturnSType, b: OperatorSType, c: OtherSType) => ReturnSType,
	): Parser<OtherSType | ReturnSType> {
		return P.sequenceMap(
			(from: ParsingPosition, first: OtherSType, others: [OperatorSType, OtherSType, ParsingPosition][]) => {
				return others.reduce<OtherSType | ReturnSType>((acc, y) => {
					const [operator, operand, to] = y;
					return combine({ from, to }, acc, operator, operand);
				}, first);
			},
			P_UTILS.position(),
			operandParser,
			P.sequence(operatorsParser.trim(this.optionalWhitespace()), operandParser, P_UTILS.position()).many(),
		);
	}

	static func<ArgsSType, ReturnSType>(
		name: string | Parser<string>,
		args: Parser<ArgsSType>,
		combine: (name: string, args: ArgsSType) => ReturnSType,
	): Parser<ReturnSType> {
		const nameParser: Parser<string> = typeof name === 'string' ? P.string(name) : name;
		return P.sequenceMap(
			(name, _lBracket, _1, args, _2, _rBracket) => {
				return combine(name, args);
			},
			nameParser,
			P.string('('),
			this.optionalWhitespace(),
			args,
			this.optionalWhitespace(),
			P.string(')'),
		);
	}
}
