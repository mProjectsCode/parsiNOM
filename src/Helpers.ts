import { type ParseResult, type ParsingPosition, type STypeBase } from './HelperTypes';
import { Parser } from './Parser';

export function arrayUnion(a: string[] | undefined, b: string[] | undefined): string[] | undefined {
	if (a === undefined && b === undefined) {
		return undefined;
	}
	if (a === undefined) {
		return b;
	}
	if (b === undefined) {
		return a;
	}

	for (const bElement of b) {
		if (!a.includes(bElement)) {
			a.push(bElement);
		}
	}
	return a;
}

export function getIndex(position: ParsingPosition | undefined): number {
	return position === undefined ? -1 : position.index;
}

export function validateRange(min: number, max: number): void {
	if (max < min) {
		throw new Error(`Invalid Range: max might not be smaller than min. Received [${min}, ${max}].`);
	}

	if (min < 0 || max < 0) {
		throw new Error(`Invalid Range: max and min might not be smaller than 0. Received [${min}, ${max}].`);
	}

	if (min === Infinity || max === Infinity) {
		throw new Error(`Invalid Range: max and min might not be infinity. Received [${min}, ${max}].`);
	}

	if (!Number.isInteger(min) || !Number.isInteger(max)) {
		throw new Error(`Invalid Range: max and min must be integers. Received [${min}, ${max}].`);
	}
}

export function validateRegexFlags(flags: string): void {
	for (const flag of flags) {
		if (flag !== 'i' && flag !== 'm' && flag !== 's' && flag !== 'u') {
			throw new Error(`RegExp flag '${flag}' is not allowed. The only allowed flags are 'i', 'm', 's' and 'u'.`);
		}
	}
}

// Parser Helpers

export class ParserHelpers {
	followedBy<SType extends STypeBase>(x: Parser<SType>): Parser<SType> {
		return new Parser<SType>(function _followedBy(context): ParseResult<SType> {
			return x.p(context.copy());
		});
	}

	/**
	 * Returns a parser that yields undefined if the input parser fails, and fails if the input parser accepts. Consumes no input.
	 * More or less an inverse lookahead.
	 *
	 * @param parser
	 */
	notFollowedBy(parser: Parser<unknown>): Parser<undefined> {
		return new Parser(function _notFollowedBy(context): ParseResult<undefined> {
			const contextCopy = context.copy();
			const result = parser.p(contextCopy);
			const text = context.sliceTo(contextCopy.position.index);
			return result.success ? context.fail(`not '` + text + `'`) : context.succeed(undefined);
		});
	}

	/**
	 * Returns a parser that passes the next character into `fn` and yields the character if `fn` returns true.
	 *
	 * @param fn
	 */
	test(fn: (char: string) => boolean): Parser<string> {
		return new Parser<string>(function _test(context): ParseResult<string> {
			const char = context.input[context.position.index];
			if (!context.atEOF() && fn(char)) {
				return context.succeedOffset(1, char);
			} else {
				return context.fail(`a character matching ${fn}`);
			}
		});
	}
}

export const P_HELPERS = new ParserHelpers();
