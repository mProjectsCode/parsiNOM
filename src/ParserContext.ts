import { arrayUnion } from './Helpers';
import type { InternalParseResult, STypeBase } from './HelperTypes';

interface LatestError {
	position: number;
	expected: string[];
}

/**
 * Holds the input string as well as the current parsing position.
 * Also provides methods for succeeding and failing a parse, as well as merging two parse results.
 */
export class ParserContext {
	readonly input: string;
	/**
	 * The current parsing position. This will be modified during parsing.
	 */
	position: number;

	latestError: LatestError | undefined;

	constructor(input: string, position: number) {
		this.input = input;
		this.position = position;
	}

	/**
	 * Checks if the parser it at or beyond the end of the input string.
	 */
	atEOF(): boolean {
		return this.position >= this.input.length;
	}

	/**
	 * Advances the parser to an index while counting lines.
	 *
	 * @param index the index to advance to, this can't be smaller than the current position
	 * @private
	 */
	private advanceTo(index: number): void {
		this.position = index;
	}

	/**
	 * Get a slice of the input from the current position to the end index.
	 *
	 * @param endIndex the index to slice to
	 */
	sliceTo(endIndex: number): string {
		return this.input.slice(this.position, endIndex);
	}

	/**
	 * Succeed at a specific offset from the current position.
	 *
	 * @param offset the offset from the current position, must be positive
	 * @param value
	 */
	succeedOffset<SType extends STypeBase>(offset: number, value: SType): InternalParseResult<SType> {
		return this.succeedAt(this.position + offset, value);
	}

	/**
	 * Fail at a specific offset from the current position.
	 *
	 * @param offset the offset from the current position, must be positive
	 * @param expected
	 */
	failOffset<SType extends STypeBase>(offset: number, expected: string | string[]): InternalParseResult<SType> {
		return this.failAt(this.position + offset, expected);
	}

	/**
	 * Succeed at the current position.
	 *
	 * @param value
	 */
	succeed<SType extends STypeBase>(value: SType): InternalParseResult<SType> {
		return this.succeedAt(this.position, value);
	}

	/**
	 * Fail at the current position.
	 *
	 * @param expected
	 */
	fail<SType extends STypeBase>(expected: string | string[]): InternalParseResult<SType> {
		return this.failAt(this.position, expected);
	}

	/**
	 * Succeed at a specific index.
	 * The index must be higher than the current position.
	 *
	 * @param index
	 * @param value
	 */
	succeedAt<SType extends STypeBase>(index: number, value: SType): InternalParseResult<SType> {
		this.advanceTo(index);

		return {
			success: true,
			value: value,
		};
	}

	/**
	 * Fail at a specific index.
	 * The index must be higher than the current position.
	 *
	 * @param index
	 * @param expected
	 */
	failAt<SType extends STypeBase>(index: number, expected: string | string[]): InternalParseResult<SType> {
		this.advanceTo(index);

		this.addError(Array.isArray(expected) ? expected : [expected]);

		return {
			success: false,
			value: undefined,
		};
	}

	addError(expected: string[]): void {
		if (this.latestError === undefined || this.latestError.position < this.position) {
			this.latestError = {
				position: this.position,
				expected: expected,
			};
		} else if (this.latestError.position === this.position) {
			this.latestError.expected = arrayUnion(this.latestError.expected, expected) ?? expected;
		}
	}

	getAndClearLatestError(): LatestError | undefined {
		const error = this.latestError;
		this.latestError = undefined;
		return error;
	}

	mergeLatestError(other: LatestError | undefined): void {
		if (other === undefined) {
			return;
		}

		if (this.latestError === undefined || this.latestError.position < other.position) {
			this.latestError = other;
		} else if (this.latestError.position === other.position) {
			this.latestError.expected = arrayUnion(this.latestError.expected, other.expected) ?? other.expected;
		}
	}
}
