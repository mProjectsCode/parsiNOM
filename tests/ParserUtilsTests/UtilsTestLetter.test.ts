import { P_UTILS } from '../../src/ParserUtils';
import { testParser } from '../TestHelpers';
import { describe } from 'bun:test';

describe.each([
	['a', true],
	['b', true],
	['z', true],
	['こ', false],
	['你', false],
	['#', false],
	[' ', false],
	['', false],
])(`letter`, (str, shouldSucceed) => {
	const parser = P_UTILS.letter().thenEof();
	testParser(parser, str, shouldSucceed);
});

describe.each([
	['a', true],
	['bb', true],
	['zz', true],
	['こんにちは', false],
	['你叫什么名字', false],
	['#', false],
	[' ', false],
	['', false],
])(`letters`, (str, shouldSucceed) => {
	const parser = P_UTILS.letters().thenEof();
	testParser(parser, str, shouldSucceed);
});

describe.each([
	['a', true],
	['b', true],
	['z', true],
	['こ', true],
	['你', true],
	['#', false],
	[' ', false],
	['', false],
])(`unicodeLetter`, (str, shouldSucceed) => {
	const parser = P_UTILS.unicodeLetter().thenEof();
	testParser(parser, str, shouldSucceed);
});

describe.each([
	['a', true],
	['bb', true],
	['zz', true],
	['こんにちは', true],
	['你叫什么名字', true],
	['#', false],
	[' ', false],
	['', false],
])(`unicodeLetters`, (str, shouldSucceed) => {
	const parser = P_UTILS.unicodeLetters().thenEof();
	testParser(parser, str, shouldSucceed);
});

describe.each([
	['a', true],
	['b', true],
	['z', true],
	['こ', true],
	['你', true],
	['1', true],
	['#', false],
	[' ', false],
	['', false],
])(`unicodeAlphanumeric`, (str, shouldSucceed) => {
	const parser = P_UTILS.unicodeAlphanumeric().thenEof();
	testParser(parser, str, shouldSucceed);
});

describe.each([
	['a', true],
	['bb', true],
	['zz', true],
	['こんにちは', true],
	['你叫什么名字', true],
	['こんにちは123', true],
	['你叫什么名字123', true],
	['#', false],
	[' ', false],
	['', false],
])(`unicodeAlphanumeric`, (str, shouldSucceed) => {
	const parser = P_UTILS.unicodeAlphanumerics().thenEof();
	testParser(parser, str, shouldSucceed);
});
