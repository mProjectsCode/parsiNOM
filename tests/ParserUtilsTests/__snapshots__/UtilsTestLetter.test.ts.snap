// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`letter 'a' conditional success tests AST matches snapshot 1`] = `"a"`;

exports[`letter 'b' conditional success tests AST matches snapshot 1`] = `"b"`;

exports[`letter 'z' conditional success tests AST matches snapshot 1`] = `"z"`;

exports[`letter 'こ' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a letter",
  ],
  "pos": 0,
}
`;

exports[`letter '你' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a letter",
  ],
  "pos": 0,
}
`;

exports[`letter '#' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a letter",
  ],
  "pos": 0,
}
`;

exports[`letter ' ' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a letter",
  ],
  "pos": 0,
}
`;

exports[`letter '' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a letter",
  ],
  "pos": 0,
}
`;

exports[`letters 'a' conditional success tests AST matches snapshot 1`] = `"a"`;

exports[`letters 'bb' conditional success tests AST matches snapshot 1`] = `"bb"`;

exports[`letters 'zz' conditional success tests AST matches snapshot 1`] = `"zz"`;

exports[`letters 'こんにちは' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "multiple letters",
  ],
  "pos": 0,
}
`;

exports[`letters '你叫什么名字' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "multiple letters",
  ],
  "pos": 0,
}
`;

exports[`letters '#' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "multiple letters",
  ],
  "pos": 0,
}
`;

exports[`letters ' ' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "multiple letters",
  ],
  "pos": 0,
}
`;

exports[`letters '' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "multiple letters",
  ],
  "pos": 0,
}
`;

exports[`unicodeLetter 'a' conditional success tests AST matches snapshot 1`] = `"a"`;

exports[`unicodeLetter 'b' conditional success tests AST matches snapshot 1`] = `"b"`;

exports[`unicodeLetter 'z' conditional success tests AST matches snapshot 1`] = `"z"`;

exports[`unicodeLetter 'こ' conditional success tests AST matches snapshot 1`] = `"こ"`;

exports[`unicodeLetter '你' conditional success tests AST matches snapshot 1`] = `"你"`;

exports[`unicodeLetter '#' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a unicode letter",
  ],
  "pos": 0,
}
`;

exports[`unicodeLetter ' ' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a unicode letter",
  ],
  "pos": 0,
}
`;

exports[`unicodeLetter '' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a unicode letter",
  ],
  "pos": 0,
}
`;

exports[`unicodeLetters 'a' conditional success tests AST matches snapshot 1`] = `"a"`;

exports[`unicodeLetters 'bb' conditional success tests AST matches snapshot 1`] = `"bb"`;

exports[`unicodeLetters 'zz' conditional success tests AST matches snapshot 1`] = `"zz"`;

exports[`unicodeLetters 'こんにちは' conditional success tests AST matches snapshot 1`] = `"こんにちは"`;

exports[`unicodeLetters '你叫什么名字' conditional success tests AST matches snapshot 1`] = `"你叫什么名字"`;

exports[`unicodeLetters '#' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "multiple unicode letters",
  ],
  "pos": 0,
}
`;

exports[`unicodeLetters ' ' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "multiple unicode letters",
  ],
  "pos": 0,
}
`;

exports[`unicodeLetters '' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "multiple unicode letters",
  ],
  "pos": 0,
}
`;

exports[`unicodeAlphanumeric 'a' conditional success tests AST matches snapshot 1`] = `"a"`;

exports[`unicodeAlphanumeric 'b' conditional success tests AST matches snapshot 1`] = `"b"`;

exports[`unicodeAlphanumeric 'z' conditional success tests AST matches snapshot 1`] = `"z"`;

exports[`unicodeAlphanumeric 'こ' conditional success tests AST matches snapshot 1`] = `"こ"`;

exports[`unicodeAlphanumeric '你' conditional success tests AST matches snapshot 1`] = `"你"`;

exports[`unicodeAlphanumeric '1' conditional success tests AST matches snapshot 1`] = `"1"`;

exports[`unicodeAlphanumeric '#' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a unicode alphanumeric character",
  ],
  "pos": 0,
}
`;

exports[`unicodeAlphanumeric ' ' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a unicode alphanumeric character",
  ],
  "pos": 0,
}
`;

exports[`unicodeAlphanumeric '' conditional failure tests Error matches snapshot 1`] = `
{
  "expected": [
    "a unicode alphanumeric character",
  ],
  "pos": 0,
}
`;

exports[`unicodeAlphanumeric 'a' conditional success tests AST matches snapshot 2`] = `"a"`;

exports[`unicodeAlphanumeric 'bb' conditional success tests AST matches snapshot 1`] = `"bb"`;

exports[`unicodeAlphanumeric 'zz' conditional success tests AST matches snapshot 1`] = `"zz"`;

exports[`unicodeAlphanumeric 'こんにちは' conditional success tests AST matches snapshot 1`] = `"こんにちは"`;

exports[`unicodeAlphanumeric '你叫什么名字' conditional success tests AST matches snapshot 1`] = `"你叫什么名字"`;

exports[`unicodeAlphanumeric 'こんにちは123' conditional success tests AST matches snapshot 1`] = `"こんにちは123"`;

exports[`unicodeAlphanumeric '你叫什么名字123' conditional success tests AST matches snapshot 1`] = `"你叫什么名字123"`;

exports[`unicodeAlphanumeric '#' conditional failure tests Error matches snapshot 2`] = `
{
  "expected": [
    "multiple unicode alphanumeric characters",
  ],
  "pos": 0,
}
`;

exports[`unicodeAlphanumeric ' ' conditional failure tests Error matches snapshot 2`] = `
{
  "expected": [
    "multiple unicode alphanumeric characters",
  ],
  "pos": 0,
}
`;

exports[`unicodeAlphanumeric '' conditional failure tests Error matches snapshot 2`] = `
{
  "expected": [
    "multiple unicode alphanumeric characters",
  ],
  "pos": 0,
}
`;
